pandas/tools/hashing.py

```diff
@@ -143,18 +145,14 @@ def hash_tuples(vals, encoding='utf8', hash_key=None):
         vals = MultiIndex.from_tuples(vals)
 
     # create a list-of-ndarrays
-    def get_level_values(num):
-        unique = vals.levels[num]  # .values
-        labels = vals.labels[num]
-        filled = algos.take_1d(unique._values, labels,
-                               fill_value=unique._na_value)
-        return filled
-
-    vals = [get_level_values(level)
+    vals = [vals._get_level_values(level)
             for level in range(vals.nlevels)]
 
     # hash the list-of-ndarrays
-    hashes = (hash_array(l, encoding=encoding, hash_key=hash_key)
+    hashes = (hash_array(l,
```

Comment:
Could be in a follow-up, but I wonder if this could be made more efficient taking advantage of the existing categorization in the `MultiIndex`?